rubix-core/src/main/java/com/qubole/rubix/core/RemoteReadRequestChain.java

```diff
@@ -61,78 +81,76 @@ public Integer call()
             }
-            if (localFile != null) {
-                localFile.close();
+
+            log.debug(String.format("Trying to Read %d bytes into destination buffer", readRequest.getActualReadLength()));
+            int readBytes = readIntoBuffer(readRequest.getDestBuffer(), readRequest.destBufferOffset, readRequest.getActualReadLength());
+            log.debug(String.format("Read %d bytes into destination buffer", readBytes));
+            copyIntoCache(readRequest.destBuffer, readRequest.destBufferOffset, readBytes, readRequest.actualReadStart);
+            log.debug(String.format("Copied %d requested bytes into cache", readBytes));
+            totalRequestedRead += readBytes;
+
+            if (suffixBufferLength > 0) {
+                // If already in reading actually required data we get a eof, then there should not have been a suffix request
+                checkState(readBytes == readRequest.getActualReadLength(), "Actual read less than required, still requested for suffix");
+                log.debug(String.format("Trying to Read %d bytes into suffix buffer", suffixBufferLength));
+                totalSuffixRead += readIntoBuffer(affixBuffer, 0, suffixBufferLength);
+                log.debug(String.format("Read %d bytes into suffix buffer", suffixBufferLength));
+                copyIntoCache(affixBuffer, 0, totalSuffixRead, readRequest.actualReadEnd);
+                log.debug(String.format("Copied %d suffix bytes into cache", suffixBufferLength));
             }
         }
         log.info(String.format("Read %d bytes from remote file, added %d to destination buffer", totalPrefixRead + totalRequestedRead + totalSuffixRead, totalRequestedRead));
         return totalRequestedRead;
     }
 
-    private int readAndCopy(byte[] destBuffer, int destBufferOffset, MappedByteBuffer localFileBuffer, int length)
+    private int readIntoBuffer(byte[] destBuffer, int destBufferOffset, int length)
             throws IOException
     {
         int nread = 0;
         while (nread < length) {
             int nbytes = inputStream.read(destBuffer, destBufferOffset + nread, length - nread);
             if (nbytes < 0) {
-                return nread;
+                break;
             }
             nread += nbytes;
         }
+        return nread;
+    }
+
+    private int copyIntoCache(byte[] destBuffer, int destBufferOffset, int length, long cacheReadStart)
+            throws IOException
+    {
+        log.debug(String.format("Trying to copy [%d - %d] bytes from destination buffer offset %d", cacheReadStart, cacheReadStart + length, destBufferOffset));
         long start = System.nanoTime();
-        localFileBuffer.put(destBuffer, destBufferOffset, length);
+        int leftToRead = length;
+        int readSoFar = 0;
+        while (leftToRead > 0) {
+            int readInThisCycle = Math.min(leftToRead, directBuffer.capacity());
```

Comment:
By read i mean reading from readrequest destination buffer. write variables corresponds to filechannel writes (to cache).